### PR TITLE
Some fixes for summary, show, and map for non-1 indices

### PR DESCRIPTION
**Pull request by timholy**
Previously I specialized this in the tests, but it's better to just make this do the right thing out of the box.

### CODE REVIEW

File: test/offsetarray.jl [hunk 1]
```diff
@@ -195,6 +194,9 @@ show(io, v)
 str = takebuf_string(io)
 show(io, parent(v))
 @test str == takebuf_string(io)
+smry = summary(v)
+@test contains(smry, "OffsetArray{Float64,1")
```
Comment (Sacha0) [inline test/offsetarray.jl]: Closing `}` omission intentional?

Comment (timholy) [inline test/offsetarray.jl]: Yes. The third parameter is the array type, and I don't care about that here.

File: base/abstractarray.jl [hunk 2]
```diff
@@ -40,7 +40,7 @@ indices{T,N}(A::AbstractArray{T,N}, d) = d <= N ? indices(A)[d] : OneTo(1)
 
 Returns the tuple of valid indices for array `A`.
 """
-function indices{T,N}(A::AbstractArray{T,N})
+function indices(A)
     @_inline_pure_meta
```
Comment (vtjnash) [inline base/abstractarray.jl]: this isn't pure

File: base/show.jl [hunk 3]
```diff
@@ -1657,17 +1660,15 @@ function show_vector(io::IO, v, opn, cls)
         io = IOContext(io, :compact => compact)
     end
     print(io, prefix)
-    if limited && length(v) > 20
-        inds = _indices1(v)
+    if limited && _length(v) > 20
```
Comment (vtjnash) [inline base/show.jl]: What is `_length`? Could we use `size(v, 1)` instead?

File: test/offsetarray.jl [hunk 4]
```diff
@@ -303,6 +310,10 @@ map!(+, dest, am, am)
 @test dest[1,8] == 4
 @test dest[1,9] == -2
 
+am = map(identity, a)
```
Comment (vtjnash) [inline test/offsetarray.jl]: wrap this in a `let` block

Comment (vtjnash) [inline test/offsetarray.jl]: hm, seems this whole file is written at global scope. that prevents gc from ever cleaning up the intermediate values. maybe we should defer this work and add this to various `@testsets` once kshyatt's PR is merged.

File: base/abstractarray.jl (hunk 2 above)
Comment (timholy) [inline base/abstractarray.jl]: Right, I should take the pure out of the fallback.

File: base/show.jl (hunk 3 above)
Comment (timholy) [inline base/show.jl]: http://docs.julialang.org/en/latest/devdocs/offset-arrays/#do-not-implement-size-or-length
http://docs.julialang.org/en/latest/devdocs/offset-arrays/#generalizing-existing-code

(This is a temporary state, we will go back to supporting `length` someday.)

File: test/offsetarray.jl (hunk 4 above)
Comment (timholy) [inline test/offsetarray.jl]: The whole file is inside a `let`: https://github.com/JuliaLang/julia/blob/c10667fda172138a376be4c72d7db3bd4b4f323e/test/offsetarray.jl#L103

File: base/abstractarray.jl (hunk 2 above)
Comment (timholy) [inline base/abstractarray.jl]: Hah, incidentally "fixed" #17126, that was unexpected.

File: test/offsetarray.jl [hunk 5]
```diff
@@ -207,6 +209,11 @@ cmp_showf(Base.print_matrix, io, OffsetArray(rand(5,5), (10,-9)))       # rows&c
 cmp_showf(Base.print_matrix, io, OffsetArray(rand(10^3,5), (10,-9)))    # columns fit
 cmp_showf(Base.print_matrix, io, OffsetArray(rand(5,10^3), (10,-9)))    # rows fit
 cmp_showf(Base.print_matrix, io, OffsetArray(rand(10^3,10^3), (10,-9))) # neither fits
+for n = 0:4
+    a = OffsetArray(rand(ntuple(d->3,n)), ntuple(identity,n))
+    show(IOContext(io, limit=true), MIME("text/plain"), a)
+    show(IOContext(io, limit=true), MIME("text/plain"), (a,a))
```
Comment (tkelman) [inline test/offsetarray.jl]: should the outputs here be tested?

**Comment by timholy** [comment]
Seem OK now?

**Comment by tkelman** [comment]
are you about to push something? should probably read from `io` and check that the outputs are as expected instead of just "anything that isn't an error is fine"

File: test/offsetarray.jl [hunk 6]
```diff
@@ -207,6 +209,23 @@ cmp_showf(Base.print_matrix, io, OffsetArray(rand(5,5), (10,-9)))       # rows&c
 cmp_showf(Base.print_matrix, io, OffsetArray(rand(10^3,5), (10,-9)))    # columns fit
 cmp_showf(Base.print_matrix, io, OffsetArray(rand(5,10^3), (10,-9)))    # rows fit
 cmp_showf(Base.print_matrix, io, OffsetArray(rand(10^3,10^3), (10,-9))) # neither fits
+targets1 = ["0-dimensional OAs.OffsetArray{Int64,0,Array{Int64,0}}:\n1",
```
Comment (tkelman) [inline test/offsetarray.jl]: will this need to be `$Int` on 32 bit systems?

**Comment by tkelman** [comment]
This passed when CI was run on the PR build, but CI failed after it was merged to master. Going to propose reverting, unless you can come up with a better fix.

**Pull request merged**